### PR TITLE
Kubernetes SD authentication options cleanup

### PR DESCRIPTION
**Pull request by jimmidyson**
/cc @fabxc @SuperQ

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -657,6 +641,22 @@ func (c *MarathonSDConfig) UnmarshalYAML(unmarshal func(interface{}) error) erro
 	return checkOverflow(c.XXX, "marathon_sd_config")
 }
 
+// KubernetesSDConfig is the configuration for Kubernetes service discovery.
+type KubernetesSDConfig struct {
+	Masters         []URL      `yaml:"masters"`
```
Comment (SuperQ) [inline config/config.go]: I found this confusing, I think the correct term here is actually API server.

File: config/config.go [hunk 2]
```diff
@@ -668,6 +668,12 @@ func (c *KubernetesSDConfig) UnmarshalYAML(unmarshal func(interface{}) error) er
 	if len(c.Masters) == 0 {
 		return fmt.Errorf("Kubernetes SD configuration requires at least one Kubernetes master")
 	}
+	if len(c.BearerToken) > 0 && len(c.BearerTokenFile) > 0 {
+		return fmt.Errorf("at most one of bearer_token & bearer_token_file must be configured")
```
Comment (SuperQ) [inline config/config.go]: This also seems incorrect.  If you are using `tls_config` with a `key_file` and `cert_file` the bearer token is unnecessary.

**Comment by SuperQ** [comment]
There are 3 current methods of authentication[0] for the kubernetes API server.  Which also extends to other components like the kubelet.

* SSL client certificates
* bearer tokens
* http basic auth

[0]: http://kubernetes.io/v1.0/docs/admin/authentication.html

File: config/config.go (hunk 2 above)
Comment (jimmidyson) [inline config/config.go]: That doesn't make this check incorrect though, right? you're saying there should be an additional validation on this section?

protocol wise, cert & bearer token auth are not necessary mutually exclusive, but you're right kubernetes implements them as separate auth options.

File: config/config.go (hunk 1 above)
Comment (jimmidyson) [inline config/config.go]: Yeah this is a legacy thing. I'll change it. Do I need to deprecate `masters ` & handle both options for a release or OK to make the breaking change? Seeing as beta only right now I'd go for breaking change?

File: config/config.go (hunk 2 above)
Comment (SuperQ) [inline config/config.go]: Oh, sorry, I miss-read the check.  nevermind, the check is correct.

File: config/config.go (hunk 1 above)
Comment (SuperQ) [inline config/config.go]: I'm all for the breaking change, but other Prometheus people may have other opinions.

Comment (fabxc) [inline config/config.go]: :+1: break it. That's why we flagged it as beta.

**Comment by jimmidyson** [comment]
@SuperQ Renamed `masters` to `api_servers` in config & relevant vars/functions.

File: documentation/examples/prometheus-kubernetes.yml [hunk 3]
```diff
@@ -18,14 +18,14 @@ scrape_configs:
...
-  - masters:
+  - api_servers:
     - 'https://kubernetes.default.svc'
     in_cluster: true
 
   relabel_configs:
   - source_labels: [__meta_kubernetes_role]
     action: keep
-    regex: ^(?:master|node)$
+    regex: ^(?:apiserver|node)$
```
Comment (fabxc) [inline documentation/examples/prometheus-kubernetes.yml]: By now our relabeling regexes are all anchored, i.e. wrapped with `^(?:<x>)$` internally.
I had a discussion with @grobie about sanity of this.

@grobie agreed to make a case for that in the next few days. For now let's remove the outer anchoring to match the current behavior, though questionable.

**Comment by fabxc** [comment]
:+1: modulo comment.
Thanks a lot for owning this integration after the initial implementation!

File: documentation/examples/prometheus-kubernetes.yml (hunk 3 above)
Comment (jimmidyson) [inline documentation/examples/prometheus-kubernetes.yml]: So switch to just `apiserver|node` in this case? Feels weird tbh. Didn't expect the non-capturing wrap in there, just the anchors.

Comment (fabxc) [inline documentation/examples/prometheus-kubernetes.yml]: Yes, you are not alone in this. Let's see where we get when revisiting this.

**Pull request merged**